### PR TITLE
[exptensions]: Add BMToR extension module

### PR DESCRIPTION
**Pull request by marian-pritsak**
Extension module is temporarily stored under experimental till the
infrastructure for integrating P4 autogenerated headers during build
will be available

Signed-off-by: Marian Pritsak <marianp@mellanox.com>

### CODE REVIEW

File: experimental/saiexperimentalbmtor.h [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiexperimentalbmtor.h
+ *
+ * @brief   This module defines SAI || P4 extension interface
```
Comment (kcudnik) [inline experimental/saiexperimentalbmtor.h]: why there are 2 pipes here? "||"

File: experimental/saiexperimentalbmtor.h [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+
+    /**
+     * @brief Action set_vnet_bitmap parameter meta_reg
+     *
+     * @type sai_uint16_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @isvlan false
+     * @condition SAI_TABLE_PEERING_ENTRY_ATTR_ACTION == SAI_TABLE_PEERING_ENTRY_ACTION_SET_VNET_BITMAP
+     */
+    SAI_TABLE_PEERING_ENTRY_ATTR_META_REG,
```
Comment (lguohan) [inline experimental/saiexperimentalbmtor.h]: since only 1 bit is enabled, can this be an index?

File: experimental/saiexperimentalbmtor.h [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_META_REG_MASK,
+
+    /**
+     * @brief Matched key dst_ip
+     *
+     * @type sai_ip_address_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_DST_IP,
```
Comment (lguohan) [inline experimental/saiexperimentalbmtor.h]: this needs to be changed to sai_ip_prefix_t

File: experimental/saiexperimentalbmtor.h [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    SAI_TABLE_VHOST_ENTRY_ATTR_PRIORITY,
+
+    /**
+     * @brief Matched key meta_reg (key)
+     *
+     * @type sai_uint16_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @isvlan false
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_META_REG_KEY,
```
Comment (lguohan) [inline experimental/saiexperimentalbmtor.h]: does this mean we can only support 16 vnet peering?

File: experimental/saiexperimentalbmtor.h [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+    SAI_TABLE_VHOST_ENTRY_ATTR_DST_IP,
+
+    /**
+     * @brief Is default entry
+     *
+     * @type bool
+     * @flags CREATE_ONLY
+     * @default false
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_IS_DEFAULT,
```
Comment (lguohan) [inline experimental/saiexperimentalbmtor.h]: what is default entry?

File: experimental/saiexperimentalbmtor.h [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_ACTION = SAI_TABLE_VHOST_ENTRY_ATTR_START,
+
+    /**
+     * @brief Rule priority in table
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_PRIORITY,
```
Comment (lguohan) [inline experimental/saiexperimentalbmtor.h]: do we need this priority if we are doing lpm here?

File: experimental/saiexperimentalbmtor.h [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+     */
+    SAI_TABLE_VHOST_ENTRY_ATTR_END,
+
+    /** Custom range base value */
+    SAI_TABLE_VHOST_ENTRY_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /** End of custom range base */
+    SAI_TABLE_VHOST_ENTRY_ATTR_CUSTOM_RANGE_END,
+
+} sai_table_vhost_entry_attr_t;
```
Comment (lguohan) [inline experimental/saiexperimentalbmtor.h]: where is the dst mac attribute?

**Comment by lguohan** [comment]
we also need crm for this resource.

File: experimental/saiexperimentalbmtor.h [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+/**
+ * @brief Clear statistics counters.
+ *
+ * @param[in] entry_id Entry id
+ * @param[in] number_of_counters Number of counters in the array
+ * @param[in] counter_ids Specifies the array of counter ids
+ *
+ * @return #SAI_STATUS_SUCCESS on success Failure status code on error
+ */
+typedef sai_status_t (*sai_clear_table_vhost_entry_stats_fn)(
```
Comment (itaibaz) [inline experimental/saiexperimentalbmtor.h]: extended statistics function is also needed

**Comment by tushar-ty** [comment]
Pls add some documentation as discussed to explain the general design. 
-thx

**Pull request merged**